common/achievement/get.ts

```diff
@@ -0,0 +1,67 @@
+import { prisma } from '../prisma';
+import { Achievement_template, User_achievement, achievement_action_type_enum, achievement_type_enum } from '../../graphql/generated';
+import { Step, Achievement, achievement_state } from '../../graphql/types/achievement';
+import { User } from '../user';
+
+const getUserAchievements = async (user: User): Promise<Achievement[]> => {
+    const userAchievements = await prisma.user_achievement.findMany({
+        where: { userId: user.userID },
+        include: { template: true },
+    });
+    const userAchievementGroups: { [group: string]: User_achievement[] } = {};
+    userAchievements.forEach((ua) => {
+        if (!userAchievementGroups[ua.template.group]) {
+            userAchievementGroups[ua.template.group] = [];
+        }
+        userAchievementGroups[ua.template.group].push(ua);
+    });
+    const achievements: Achievement[] = createFrontendAchievements(userAchievementGroups);
+    return achievements;
+};
+
+const createFrontendAchievements = (groups: { [group: string]: User_achievement[] }): Achievement[] => {
+    const achievements: Achievement[] = [];
+    for (const group in groups) {
+        const sortedGroupAchievements = groups[group].sort((a, b) => a.groupOrder - b.groupOrder);
+        const currentAchievementIndex = sortedGroupAchievements.findIndex((ua) => !ua.achievedAt) - 1;
+        const state: achievement_state =
+            sortedGroupAchievements.length === 0
```

Comment:
I still don't get how `sortedGroupAchievements.length === 0` can happen 😅 

In lines 12 - 17, we only create a group if there is an achievement, and it's pushed to the array immediately. This means that if we have a group, it has at least one achievement in any case.

It would only make sense if we somehow try to compare templates with user_achievements.

Reply:
Ok this is a correct thought. To make this reasonable we'll have to make this userAchievementGroups map to be created from templates and not already existing achievements. I'll think about how to fix this!